src/factories/implementations/dependency-injection-factory.php

```diff
@@ -0,0 +1,138 @@
+    $this->method_name = $method_name;
+    $this->arguments   = $arguments;
+    $this->validators  = [];
+  }
+
+  /**
+   * Build the object, its dependencies, and list
+   * any validators that should be checked
+   *
+   * @override
+   */
+  public function build() {
+    $reflector = new \ReflectionMethod(
+        $this->class_name,
+        $this->method_name
+    );
+
+    $dependencies = $this->get_dependencies(
+        $reflector,
+        $this->arguments
+    );
+
+    $obj = $this->make(
+        $this->class_name,
+        $this->arguments
+    );
+
+    return [ $obj, $dependencies, $this->validators ];
+  }
+
+  /**
+   * Get the dependencies for a given reflection object
+   *
+   * @param $reflector ReflectionMethod Reflection object to gather dependencies from
+   * @param $named_arguments Array An associative array of suggested arguments
+   */
+  private function get_dependencies( \ReflectionMethod $reflector, $named_arguments ) {
+    $reflector_parameters = $reflector->getParameters();
+
+    $dependencies = $this->resolve_dependencies( $reflector_parameters, $named_arguments );
+
+    return $dependencies;
+  }
+
+  /**
+   * Resolve all dependencies recursively. Untyped parameters will
+   * have the $named_arguments injected into them
+   *
+   * @param $reflector_parameters Array<ReflectionParameters> Will check these for dependencies
+   * @param $named_arguments Array An associative array of suggested arguments
+   */
+  private function resolve_dependencies( $reflector_parameters, $named_arguments = [] ) {
+    // map the named_arguments to the reflector_parameters
+    $dependencies = [];
+
+    foreach ( $reflector_parameters as $parameters ) {
+      if ( array_key_exists( $parameters->getName(), $named_arguments ) ) {
+        $dependencies[] = $named_arguments[ $parameters->getName() ];
+
+        // TODO Need to determine whether used parameters should
+        // be popped. If they should, uncomment out the following
+        // line.
+        // unset( $named_arguments[ $parameters->getName() ] );
+      } else if ( $parameters->getClass() ) {
+        $dependencies[] = $this->make( $parameters->getClass()->name, $named_arguments );
+      } else if ( $parameters->isDefaultValueAvailable() ) {
+        $dependencies[] = $parameters->getDefaultValue();
+      } else {
+        $dependencies[] = $named_arguments;
+      }
+      // TODO inject default values (if available)
+    }
+    return $dependencies;
+  }
+
+  /**
+   * Given a class and arguments to inject into that class, this
+   * will create an instance of the given object.
+   *
+   * This will handle special cases as well.
+   *
+   * @param $class_name String The name of the class to make
+   * @param $named_arguments Array The arguments to inject into the class
+   * @return Mixed An instance of the class
+   */
+  private function make( $class_name, $named_arguments ) {
+    $reflector   = new \ReflectionClass( $class_name );
```

Comment:
I had to remove a special case where we made a Gios_Api\PDO object. I think we just need to change the PDO class to have Dependency Injection and have the Gios_Api be a singleton.